htmd/parameterization/detect.py

```diff
@@ -0,0 +1,366 @@
+    Terminal cores are skipped.
+    """
+
+    methyl = getMethylGraph()
+
+    all_core_sides = []
+    for core in list(nx.bridges(graph)):
+
+        # Get side graphs of the core
+        graph.remove_edge(*core)
+        sideGraphs = list(nx.connected_component_subgraphs(graph))
+        graph.add_edge(*core)
+
+        # Skip terminal bridges, which cannot form dihedral angles
+        if len(sideGraphs[0]) == 1 or len(sideGraphs[1]) == 1:
+            continue
+
+        # Swap the side graphs to match the order of the core
+        sideGraphs = sideGraphs[::-1] if core[0] in sideGraphs[1] else sideGraphs
+        assert core[0] in sideGraphs[0] and core[1] in sideGraphs[1]
+
+        # Skip if a side graph is a methyl group
+        if checkIsomorphism(sideGraphs[0], methyl) or checkIsomorphism(sideGraphs[1], methyl):
+            continue
+
+        # Skip if core contains C with sp hybridization
+        if graph.nodes[core[0]]['element'] == 'C' and graph.degree(core[0]) == 2:
+            continue
+        if graph.nodes[core[1]]['element'] == 'C' and graph.degree(core[1]) == 2:
+            continue
+
+        all_core_sides.append((core, sideGraphs))
+
+    return all_core_sides
+
+def weighted_closeness_centrality(graph, node, weight=None):
+    '''
+    Weighted closeness centrality
+
+    Identical to networkx.closeness_centrality, except the shorted path lengths are weighted by a node attribute.
+    '''
+
+    lengths = nx.shortest_path_length(graph, source=node)
+    del lengths[node]
+    weights = {node_: graph.nodes[node_][weight] for node_ in lengths} if weight else {node_: 1 for node_ in lengths}
+    centrality = sum(weights.values())/sum([lengths[node_]*weights[node_] for node_ in lengths])
+
+    return centrality
+
+def chooseTerminals(graph, centre, sideGraph):
+    """
+    Choose dihedral angle terminals (outer atoms)
+
+    The terminals are chosen by:
+    1. Largest closeness centrality
+    2. Largest atomic number weighted closeness centrality
+    """
+
+    terminals = list(sideGraph.neighbors(centre))
+
+    # Get a subgraph for each terminal
+    sideGraph = sideGraph.copy()
+    sideGraph.remove_node(centre)
+    terminalGraphs = list(nx.connected_component_subgraphs(sideGraph))
+    terminalGraphs = [[terminalGraph for terminalGraph in terminalGraphs if terminal in terminalGraph.nodes][0] for terminal in terminals]
+
+    # Compute a score for each terminal
+    centralities = [nx.closeness_centrality(graph, terminal) for terminal in terminals]
+    weightedCentralities = [weighted_closeness_centrality(graph, terminal, weight='number') for terminal in terminals]
+    scores = list(zip(centralities, weightedCentralities))
+
+    # Choose the terminals
+    chosen_terminals = []
+    refTerminalGraph = None
+    for terminal, score, terminalGraph in zip(terminals, scores, terminalGraphs):
+        if score < max(scores):
+            continue
+
```

Comment:
So you don't really want a loop. Do an `argmax` on `scores` and index `terminals` and `terminalGraphs` with that value.

Reply:
Usually there is more than one terminal with the same score, so `argmax` isn't a replacement here.